### PR TITLE
Add config model

### PR DESCRIPTION
**Pull request by henrikstranneheim**
## Description

Add host of analysis to config and as a environment variable for the server to make TB configurable.

### Added

- ANALYSIS_HOST to dockerfile to enable setting it at runtime
- Add and use Pydantic Config model for CLI input
- Tests

### Changed

- Use `trailblazer_db` instead of  ´trailblazer` in click.context
- Remove `database` from click.context as it was not used
- Remove `database` CLI option
- Refactor `database` in config to `database_url`
- Require config option on CLI
- Use ANALYSIS_HOST when using SSH

### How to prepare for test
- [ ] ssh to hasta (depending on type of change)
- [ ] activate stage: `us`
- [ ] request trailblazer-stage on hasta: `paxa`
- [ ] install on stage:
    ```Shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_trailblazer -t trailblazer -b add_config_model -a
    ```
- [ ] ssh to clinical-db (depending on type of change)
- [ ] install on stage:
`bash /home/proj/production/servers/resources/clinical-db.scilifelab.se/update-trailblazer-ui-stage.sh [THIS-BRANCH-NAME]`

### How to test
- [ ] login to ...
- [ ] do ...

### Expected test outcome
- [ ] check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review
- [ ] tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [x] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

**Comment by henrikstranneheim** [comment]
### CLI and API ✅ 

Screenshot 2023-09-05 at 16 34 35

```
Action running set for case justhusky
mip-dna run started!
θ70° [henrik.stranneheim@hasta:~] [S_base] 1m36s $ trailblazer scan
Updated status justhusky - 278304: running
Updated status uphippo - 278303: running
Updated status equalbug - 278302: error
Updated status setamoeba - 278301: running
Updated status civilsole - 278299: running
Error updating analysis for: case - finerrhino : PermissionError
Error updating analysis for: case - immuneamoeba : PermissionError
Error updating analysis for: case - modestakita : PermissionError
Error updating analysis for: case - famouscorgi : PermissionError
Error updating analysis for: case - topskite : PermissionError
Error updating analysis for: case - firstviper : PermissionError
Error updating analysis for: case - equalthrush : PermissionError
Error updating analysis for: case - openduck : PermissionError
Error updating analysis for: case - busyturkey : PermissionError
Error updating analysis for: case - strongsponge : PermissionError
Error updating analysis for: case - helpedegret : PermissionError
Error updating analysis for: case - nativeloon : PermissionError
Error updating analysis for: case - supercougar : PermissionError
Error updating analysis for: case - socialcrab : PermissionError
Error updating analysis for: case - activelion : PermissionError
Error updating analysis for: case - amusedspider : PermissionError
Error updating analysis for: case - greatbear : PermissionError
Error updating analysis for: case - aceviper : PermissionError
Error updating analysis for: case - brightape : PermissionError
Error updating analysis for: case - wittyghoul : PermissionError
Error updating analysis for: case - safeshiner : PermissionError
Error updating analysis for: case - brightegret : PermissionError
Error updating analysis for: case - fitloon : PermissionError
Error updating analysis for: case - decentcalf : PermissionError
Error updating analysis for: case - knownburro : PermissionError
Error updating analysis for: case - smoothlocust : PermissionError
Error updating analysis for: case - cuteprawn : PermissionError
Error updating analysis for: case - frankpigeon : PermissionError
Error updating analysis for: case - politeowl : PermissionError
Error updating analysis for: case - likedmutt : PermissionError
Error updating analysis for: case - wholegull : PermissionError
Error updating analysis for: case - promptowl : PermissionError
Error updating analysis for: case - nicewhale : PermissionError
Error updating analysis for: case - activedodo : PermissionError
Error updating analysis for: case - guidedmole : PermissionError
Error updating analysis for: case - busyturtle : PermissionError
Error updating analysis for: case - caringdrum : PermissionError
Error updating analysis for: case - lovedcaiman : PermissionError
Error updating analysis for: case - awakemarten : PermissionError
Error updating analysis for: case - neededbass : PermissionError
Error updating analysis for: case - divinesawfly : PermissionError
Error updating analysis for: case - modestoriole : PermissionError
Error updating analysis for: case - onekoi : PermissionError
Error updating analysis for: case - immunefawn : PermissionError
Error updating analysis for: case - manycow : PermissionError
Error updating analysis for: case - heroicmule : PermissionError
Error updating analysis for: case - inorca : PermissionError
Error updating analysis for: case - protahr : PermissionError
Error updating analysis for: case - trustysquid : PermissionError
Error updating analysis for: case - solidhusky : PermissionError
Error updating analysis for: case - moralbeagle : PermissionError
Error updating analysis for: case - cracktiger : PermissionError
Error updating analysis for: case - daringrat : PermissionError
Error updating analysis for: case - nativepossum : PermissionError
Error updating analysis for: case - savingstag : FileNotFoundError
Error updating analysis for: case - HW2V2DRX2 : FileNotFoundError
Error updating analysis for: case - divinetomcat : PermissionError
Error updating analysis for: case - handyshrimp : FileNotFoundError
Error updating analysis for: case - setcolt : PermissionError
Error updating analysis for: case - proudfish : PermissionError
Error updating analysis for: case - funnyaphid : FileNotFoundError
Error updating analysis for: case - onestag : PermissionError
Error updating analysis for: case - newgopher : PermissionError
Error updating analysis for: case - helpedlab : MissingSqueueOutput
Error updating analysis for: case - goodbengal : MissingSqueueOutput
Error updating analysis for: case - grandaphid : PermissionError
All analyses updated!
```
Screenshot 2023-09-05 at 16 38 41
Screenshot 2023-09-05 at 16 40 53

```
5° [henrik.stranneheim@hasta:~] [S_base] 8s 2 $ trailblazer cancel 278304
Cancelling job 5094902 - bwa_mem2
Cancelling job 5094901 - bwa_mem2
Cancelling job 5094900 - bwa_mem2
Cancelling job 5094903 - samtools_merge
Cancelling job 5094904 - samtools_merge
Cancelling job 5094905 - samtools_merge
Cancelling job 5094906 - markduplicates
Cancelling job 5094907 - markduplicates
Cancelling job 5094908 - markduplicates
Cancelling job 5094934 - cnvnator_ar
Cancelling job 5094935 - cnvnator_ar
Cancelling job 5094936 - cnvnator_ar
Cancelling job 5094938 - manta
Cancelling job 5094942 - sv_varianteffectpredictor
Cancelling job 5094943 - sv_vcfparser
Cancelling job 5094944 - sv_rankvariant
Cancelling job 5094957 - deepvariant
Cancelling job 5094958 - deepvariant
Cancelling job 5094959 - deepvariant
Cancelling job 5094960 - glnexus_merge
Cancelling job 5094976 - prepareforvariantannotationblock
Cancelling job 5094977 - rhocall_ar
Cancelling job 5094978 - bcftools_norm
Cancelling job 5094979 - variant_annotation
Cancelling job 5094989 - frequency_filter
Cancelling job 5094990 - cadd_ar
Cancelling job 5094991 - varianteffectpredictor
Cancelling job 5094992 - vcfparser_ar
Cancelling job 5094993 - rankvariant
Cancelling job 5094909 - chanjo_sexcheck
Cancelling job 5094910 - chanjo_sexcheck
Cancelling job 5094911 - chanjo_sexcheck
Cancelling job 5094912 - samtools_subsample_mt
Cancelling job 5094913 - samtools_subsample_mt
Cancelling job 5094914 - samtools_subsample_mt
Cancelling job 5094915 - tiddit_coverage
Cancelling job 5094916 - tiddit_coverage
Cancelling job 5094917 - tiddit_coverage
Cancelling job 5094918 - chromograph_cov
Cancelling job 5094919 - chromograph_cov
Cancelling job 5094920 - chromograph_cov
Cancelling job 5094921 - mitodel
Cancelling job 5094922 - mitodel
Cancelling job 5094923 - mitodel
Cancelling job 5094924 - sambamba_depth
Cancelling job 5094925 - sambamba_depth
Cancelling job 5094926 - sambamba_depth
Cancelling job 5094927 - smncopynumbercaller
Cancelling job 5094928 - picardtools_collectmultiplemetrics
Cancelling job 5094929 - picardtools_collectmultiplemetrics
Cancelling job 5094930 - picardtools_collectmultiplemetrics
Cancelling job 5094931 - picardtools_collecthsmetrics
Cancelling job 5094932 - picardtools_collecthsmetrics
Cancelling job 5094933 - picardtools_collecthsmetrics
Cancelling job 5094937 - expansionhunter
Cancelling job 5094939 - tiddit
Cancelling job 5094940 - sv_combinevariantcallsets
Cancelling job 5094941 - sv_annotate
Cancelling job 5094945 - sv_reformat
Cancelling job 5094946 - vcf2cytosure_ar
Cancelling job 5094947 - me_merge_bam
Cancelling job 5094948 - me_merge_bam
Cancelling job 5094949 - me_merge_bam
Cancelling job 5094950 - retroseq
Cancelling job 5094951 - retroseq
Cancelling job 5094952 - retroseq
Cancelling job 5094953 - me_merge_vcfs
Cancelling job 5094954 - me_annotate
Cancelling job 5094955 - me_varianteffectpredictor
Cancelling job 5094956 - me_filter
Cancelling job 5094961 - gatk_combinevariantcallsets
Cancelling job 5094962 - peddy_ar
Cancelling job 5094963 - plink
Cancelling job 5094964 - gatk_variantevalall
Cancelling job 5094965 - gatk_variantevalall
Cancelling job 5094966 - gatk_variantevalall
Cancelling job 5094967 - gatk_collectreadcounts
Cancelling job 5094968 - gatk_collectreadcounts
Cancelling job 5094969 - gatk_collectreadcounts
Cancelling job 5094970 - gatk_denoisereadcounts
Cancelling job 5094971 - gatk_denoisereadcounts
Cancelling job 5094972 - gatk_denoisereadcounts
Cancelling job 5094973 - gens_generatedata
Cancelling job 5094974 - gens_generatedata
Cancelling job 5094975 - gens_generatedata
Cancelling job 5094980 - rhocall_viz
Cancelling job 5094981 - rhocall_viz
Cancelling job 5094982 - rhocall_viz
Cancelling job 5094983 - chromograph_rhoviz
Cancelling job 5094984 - chromograph_rhoviz
Cancelling job 5094985 - chromograph_rhoviz
Cancelling job 5094986 - upd_ar
Cancelling job 5094987 - chromograph_upd
Cancelling job 5094988 - mt_annotation
Cancelling job 5094994 - endvariantannotationblock
Cancelling job 5094995 - gatk_variantevalexome
Cancelling job 5094996 - gatk_variantevalexome
Cancelling job 5094997 - gatk_variantevalexome
Cancelling job 5094998 - qccollect_ar
Cancelling job 5094999 - version_collect_ar
Cancelling job 5095000 - multiqc_ar
Cancelling job 5095001 - analysisrunstatus
Cancelling job 5095002 - sacct
Cancelling job 5094899 - fastqc_ar
Cancelling job 5094897 - fastqc_ar
Cancelling job 5094898 - fastqc_ar
Case justhusky - Analysis 278304: all ongoing jobs cancelled successfully!
```
Screenshot 2023-09-05 at 16 44 56

```
θ64° [henrik.stranneheim@hasta:~] [S_base] 11s $ trailblazer delete 278304
Deleting analysis 278304 for case justhusky
```

File: tests/test_models.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+from trailblazer.models import Config
+
+
+def test_instantiate_config():
```
Comment (diitaz93) [inline tests/test_models.py:4]: Would adding a test instantiating a config without the `analysis_host` be helpful?

**Review by diitaz93 — APPROVED** [review]
Looks good 👍

**Review by Vince-janv — COMMENTED** [review]
Nice to remove hard-coded strings! I'm  not sure I like both the `use_ssh` and `analysis_host` spreading through functions where they are not used. I have some suggestions for simplifying 👍

File: tests/test_models.py (hunk 1 above)
Comment (henrikstranneheim) [reply tests/test_models.py:4]: Not needed snymore

File: trailblazer/cli/core.py [hunk 2]
```diff
@@ -169,8 +170,13 @@ def unarchive_user(context, email: str) -> None:
...
 def cancel(context, analysis_id):
     """Cancel all jobs in a run."""
+    trailblazer_db: Store = context.obj["trailblazer_db"]
     try:
-        context.obj["trailblazer"].cancel_analysis(analysis_id=analysis_id, email=environ_email())
+        trailblazer_db.cancel_analysis(
+            analysis_id=analysis_id,
+            analysis_host=context.obj["analysis_host"],
+            email=environ_email(),
+        )
```
[the remark this answers is not in the document]
Comment (henrikstranneheim) [reply trailblazer/cli/core.py]: Good suggestion - fixed

**Comment by henrikstranneheim** [comment]
@diitaz93 @Vince-janv Ready for another review.

File: trailblazer/store/api.py [hunk 3]
```diff
@@ -198,23 +202,23 @@ def update_tower_run_status(self, analysis_id: int) -> None:
             analysis.status: str = TrailblazerStatus.ERROR
             self.commit()
 
-    def cancel_analysis(self, analysis_id: int, email: str = None, ssh: bool = False) -> None:
+    def cancel_analysis(
+        self, analysis_id: int, analysis_host: Optional[str] = None, email: str = None
```
[the remark this answers is not in the document]
Comment (henrikstranneheim) [reply trailblazer/store/api.py]: Fixed

File: trailblazer/apps/slurm/api.py [hunk 4]
```diff
@@ -22,17 +22,15 @@ def _get_squeue_jobs_flag_input(slurm_job_id_file_content: Dict[str, List[str]])
...
-        scancel_commands = ["ssh", "hiseq.clinical@hasta.scilifelab.se"] + scancel_commands
-        subprocess.Popen(scancel_commands)
-        return
+    if analysis_host:
+        scancel_commands = ["ssh", analysis_host] + scancel_commands
     subprocess.Popen(scancel_commands)
 
 
-def get_slurm_squeue_output(slurm_job_id_file: Path, use_ssh: bool = False) -> str:
+def get_slurm_squeue_output(slurm_job_id_file: Path, analysis_host: str = None) -> str:
```
[the remark this answers is not in the document]
Comment (henrikstranneheim) [reply trailblazer/apps/slurm/api.py]: Fixed

File: trailblazer/store/api.py [hunk 5]
```diff
@@ -147,13 +147,17 @@ def update_run_status(self, analysis_id: int, use_ssh: bool = False) -> None:
         if analysis.workflow_manager == WorkflowManager.TOWER.value:
             self.update_tower_run_status(analysis_id=analysis_id)
         elif analysis.workflow_manager == WorkflowManager.SLURM.value:
-            self.update_analysis_from_slurm_output(analysis_id=analysis_id, use_ssh=use_ssh)
+            self.update_analysis_from_slurm_output(
+                analysis_id=analysis_id, analysis_host=analysis_host
+            )
 
-    def update_analysis_from_slurm_output(self, analysis_id: int, use_ssh: bool = False) -> None:
+    def update_analysis_from_slurm_output(self, analysis_id: int, analysis_host: str) -> None:
```
[the remark this answers is not in the document]
Comment (henrikstranneheim) [reply trailblazer/store/api.py]: Fixed

File: trailblazer/apps/slurm/api.py [hunk 6]
```diff
@@ -22,17 +22,15 @@ def _get_squeue_jobs_flag_input(slurm_job_id_file_content: Dict[str, List[str]])
     return ",".join(job_ids)
 
 
-def cancel_slurm_job(slurm_id: int, use_ssh: bool = False) -> None:
+def cancel_slurm_job(slurm_id: int, analysis_host: str = None) -> None:
```
[the remark this answers is not in the document]
Comment (henrikstranneheim) [reply trailblazer/apps/slurm/api.py]: Fixed

**Comment by henrikstranneheim** [comment]
Deployed to stage and production

**Pull request merged**